### PR TITLE
testing NN on Mendota

### PR DESCRIPTION
**Pull request by wdwatkins**
Just some scripts testing out a neural net on Mendota.  These files can be reused and eventually removed when we get to building out the actual pipeline.

`in` contains the files @jread-usgs uploaded to Jared's repo.  I reran that simulation to get access to the raw netcdf output (not included here due to size).  It also has the full observation record from `pgml-temperature-prediction`.

The script implements a basic three hidden layer neural network with keras/tensorflow, following the parameters descriped in Anuj's paper.  I experimented with some of them (mostly regularization values).  The script also makes some plots of the predictions.  I found it easier to use ggplot for heatmaps than the `glmtools` plots, since the `glmtools` functions were written to accept files.

### CODE REVIEW

File: src/functions.R [hunk 1]
```diff
@@ -0,0 +1,50 @@
+temp_to_density <- function(x) {
```
Comment (jread-usgs) [inline src/functions.R]: maybe a good place to use `rLakeAnalyzer::water.density()`

File: src/functions.R [hunk 2]
```diff
@@ -0,0 +1,46 @@
+identify_out_of_order <- function(vec){
```
Comment (jread-usgs) [inline src/functions.R]: Can you add a comment for the purpose of this function? It is different than the return of 
```r
vec == cummin(vec)
```
which tests for monotonically decreasing.

File: src/test_nn.R [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+source('src/functions.R')
+obs <- readRDS('in/nhd_13293262.rds')
+
+data.table::fwrite(obs, file = "in/mendota_obs.csv")
+meteo <- feather::read_feather('in/Mendota_meteo.feather') %>% mutate(date = as.Date(time))
+temps <- glmtools::resample_to_field(nc_file = "in/output.nc", field_file = 'in/mendota_obs.csv') %>% 
+  mutate(date = as.Date(DateTime)) %>% filter(!is.na(Modeled_temp)) 
+#NAs in input result in NaNs from loss function
+
+all_model_temps_short <- glmtools::get_temp('in/output.nc',  reference = "surface", z_out = seq(0,25,0.5)) %>%
```
Comment (jread-usgs) [inline src/test_nn.R]: Is this different than `in/Mendota_simulation.feather`?

File: src/test_nn.R [hunk 4]
```diff
@@ -0,0 +1,131 @@
+rm(list=ls())
```
Comment (jread-usgs) [inline src/test_nn.R]: I can't comment on the binaries, but isn't the shortwave data already in the `meteo` feather? Why is that needed + `in/Mendota_shortwave.feather` ?

File: src/test_nn.R [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+all_model_temps <- all_model_temps_short %>% 
+  tidyr::gather(key = "Depth", value = "Modeled_temp", -DateTime) %>% 
+  mutate(Depth = as.numeric(gsub(pattern = "temp_", replacement = "", x = Depth))) %>% 
+  filter(!is.na(Modeled_temp))
+  
+all_model_temps_drivers <- left_join(all_model_temps, meteo, by = c(DateTime = "date")) %>% 
+  augment_features()
+
+#all data, adding yday, growing deg days, snow to binary, binary freezing
+temps_meteo_joined <- dplyr::left_join(temps, meteo, by = "date") %>% augment_features() %>%
```
Comment (jread-usgs) [inline src/test_nn.R]: the naming of this is kind of confusing. It seems redundant w/ `all_model_temps_drivers`, but this is the temperature data sampled down to the time and location of the observations, right?

File: src/test_nn.R [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+              activation = "relu") %>%
+  layer_dense(units = 12, kernel_regularizer = regularizer_l1_l2(l1_lambda, l2_lambda),
+              activation = "relu", kernel_initializer = initializer_random_uniform(0,1)) %>%
+  layer_dense(units = 1, kernel_regularizer = regularizer_l1_l2(l1_lambda, l2_lambda),
+              kernel_initializer = initializer_random_uniform(0,1))
+summary(network)
+network %>% compile(
+  optimizer = optimizer_adadelta(clipnorm = 1),
+  loss = "mean_squared_error",
+  metrics = "mean_squared_error") #keras objects are modified in place!
```
Comment (jread-usgs) [inline src/test_nn.R]: Are they python objects? or `R6`?

File: src/test_nn.R [hunk 7]
```diff
@@ -0,0 +1,131 @@
+rm(list=ls())
+
+library(dplyr)
```
Comment (jread-usgs) [inline src/test_nn.R]: Again, can't comment on the binaries, so I will do it here. Seems there are a bunch of binaries that aren't used that are checked in, and another binary that is used (`output.nc`) that isn't checked in. I get that `output.nc` is likely too big, but including a script/snippet that generates that file seems appropriate. And leaving the unused files out would be good too. Otherwise, reproducibility is a challenge.

File: src/functions.R [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  }
+  return(keep)
+}
+
+augment_features <- function(df){
+  df %>% select(-time) %>% 
+    mutate(yday = lubridate::yday(DateTime),
+           Snow = ifelse(Snow > 0, yes = 1, no = 0),
+           freezing = ifelse(AirTemp <= 0, yes = 1, no = 0),
+           gdd = AirTemp - 10, gdd = ifelse(gdd < 0, yes = 0, no = gdd
```
[the remark this answers is not in the document]
Comment (wdwatkins) [reply src/functions.R]: I did wonder what just subtracting 10 from the temperature would accomplish... that would reset after winter then?

File: src/test_nn.R (hunk 6 above)
Comment (wdwatkins) [reply src/test_nn.R]: I think python:

File: src/test_nn.R [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+network %>% compile(
+  optimizer = optimizer_adadelta(clipnorm = 1),
+  loss = "mean_squared_error",
+  metrics = "mean_squared_error") #keras objects are modified in place!
+
+#fitting 
+#paper: 10k max epochs, patience 500
+history <- network %>% fit(train_data, train_targets, epochs = 5000, batch_size = 1000,
+                callbacks = callback_early_stopping(patience = 500, monitor = "loss"),
+                validation_split = 0.1)
```
Comment (aappling-usgs) [inline src/test_nn.R]: How long did this take? did you use CPU with 1 core, or is there parallelization somewhere in here already?

File: src/test_nn.R (hunk 7 above)
Comment (wdwatkins) [reply src/test_nn.R]: Right, I will get rid of the feathers I didn't end up using.  

As far as generating `output.nc`, I just reran the remake file that @jread-usgs checked in to Jared's repo.  Is it worth reposting that here?  @aappling-usgs I can just slack that to you if you want if you don't have access

File: src/test_nn.R (hunk 4 above)
Comment (wdwatkins) [reply src/test_nn.R]: it isn't needed

File: src/test_nn.R (hunk 3 above)
Comment (wdwatkins) [reply src/test_nn.R]: Shouldn't be, it was just easier to have the nc file to use the `glmtools` functions

File: src/functions.R (hunk 2 above)
Comment (wdwatkins) [reply src/functions.R]: Good point @aappling-usgs.  I believe Anuj filtered out physically inconsistent observations, so I was trying to do the same to see if it improved performance at all (it didn't, at least the way I implemented it).  

Will add some comments to that effect.

File: src/test_nn.R (hunk 7 above)
Comment (aappling-usgs) [reply src/test_nn.R]: Slack would work for me, thanks

File: src/functions.R (hunk 2 above)
Comment (wdwatkins) [reply src/functions.R]: It is different from `cummin()` in that `cummin` throws out everything after a deviation from monotonically increasing.  When this function encounters a point that is not monotonic, it discards it, but then compares to the next one.  It was an attempt to not throw out entire temp profiles based on a single bad observation.

Comment (aappling-usgs) [reply src/functions.R]: hmm, but does vec == cummin(vec) really throw away the whole profile?
```
> vec = c(9:5,3,4,5:1)
> vec
 [1] 9 8 7 6 5 3 4 5 4 3 2 1
> vec == cummin(vec)
 [1]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE  TRUE  TRUE  TRUE
> vec[vec==cummin(vec)]
[1] 9 8 7 6 5 3 3 2 1
```

File: src/test_nn.R (hunk 7 above)
Comment (wdwatkins) [reply src/test_nn.R]: Emailed it instead, it was a biggish directory

File: src/functions.R (hunk 2 above)
Comment (wdwatkins) [reply src/functions.R]: Sorry, I meant it will throw away the whole profile if the surface observation is bad (since it throws out everything after the deviation from monotonic)

File: src/functions.R [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  glmtools:::.plot_df_heatmap(glm_temp, bar_title="GLM", xaxis = xaxis, col_lim = clim)
+  glmtools:::.plot_df_heatmap(nn_temp, bar_title="NN", xaxis = xaxis, col_lim = clim)
+  glmtools:::.plot_df_heatmap(diff_temp, bar_title="NN - GLM", xaxis = xaxis, col_lim = clim)
+}
+
+
+#downsample dates to approx nobs from paper - only keep every fourth date
+# dates_keep <- unique(obs$DateTime)[c(TRUE,FALSE,FALSE,FALSE)]
+# obs <- obs %>% filter(DateTime %in% dates_keep)
+#TODO: eliminate obs that are physically inconsistent?
```
[the remark this answers is not in the document]
Comment (wdwatkins) [reply src/functions.R]: Yeah, it is probably worth having some discussion around.

File: src/functions.R (hunk 2 above)
Comment (jread-usgs) [reply src/functions.R]: how do you know the out of place one is really the bad obs? couldn't it be the one before it and/or after it instead?

File: src/test_nn.R [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+              kernel_regularizer = regularizer_l1_l2(l1_lambda, l2_lambda),
+              kernel_initializer = initializer_random_uniform(0,1),
+              activation = "relu") %>%
+  layer_dense(units = 12, kernel_regularizer = regularizer_l1_l2(l1_lambda, l2_lambda),
+              kernel_initializer = initializer_random_uniform(0,1),
+              activation = "relu") %>%
+  layer_dense(units = 12, kernel_regularizer = regularizer_l1_l2(l1_lambda, l2_lambda),
+              activation = "relu", kernel_initializer = initializer_random_uniform(0,1)) %>%
+  layer_dense(units = 1, kernel_regularizer = regularizer_l1_l2(l1_lambda, l2_lambda),
+              kernel_initializer = initializer_random_uniform(0,1))
```
[the remark this answers is not in the document]
Comment (wdwatkins) [reply src/test_nn.R]: Let continue this in #5

[the remark this answers is not in the document]
Comment (wdwatkins) [reply src/test_nn.R]: I didn't experiment with the network depth at all --- I assume Anuj did a fair bit, and this was the winner

File: src/test_nn.R [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+#all data, adding yday, growing deg days, snow to binary, binary freezing
+temps_meteo_joined <- dplyr::left_join(temps, meteo, by = "date") %>% augment_features() %>% 
+  select(-DateTime)
+n_obs <- nrow(temps_meteo_joined)
+
+#training on middle third for now
+#median_date <- median(temps_meteo_joined$date)
+training <- slice(temps_meteo_joined, (n_obs/3):(2/3*n_obs)) 
+test <- anti_join(temps_meteo_joined, training)
```
[the remark this answers is not in the document]
Comment (wdwatkins) [reply src/test_nn.R]: Not sure what you mean here, can you elaborate?

File: src/test_nn.R [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+#training on middle third for now
+#median_date <- median(temps_meteo_joined$date)
+training <- slice(temps_meteo_joined, (n_obs/3):(2/3*n_obs)) 
+test <- anti_join(temps_meteo_joined, training) 
+
+train_targets <- training$Observed_temp
+train_dates <- training$date
+test_targets <- test$Observed_temp
+test_dates <- test$date
```
[the remark this answers is not in the document]
Comment (wdwatkins) [reply src/test_nn.R]: Fair point.  You do generally have to remove the observation column from the training data at some point before feeding it to the model though.

File: src/test_nn.R [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+test_data <- as.matrix(select(test, -Observed_temp, -date)) %>% 
+  scale(center = attr(train_data, "scaled:center"), 
+        scale = attr(train_data, "scaled:scale"))
+all_model_unlabeled_data <- as.matrix(select(all_model_temps_drivers, -DateTime)) %>% 
+  scale(center = attr(train_data, "scaled:center"), 
+        scale = attr(train_data, "scaled:scale"))
+
+#build the neural net
+l2_lambda <- 0.5 #regularization factors, L1 and L2 regularization
+l1_lambda <- 0.5
```
[the remark this answers is not in the document]
Comment (wdwatkins) [reply src/test_nn.R]: I haven't thought much about it.  I did try varying the regularization weights, but kept l1 == l2.  Overfitting gets worse with lower regularization, so some form is definitely needed.  But it would be easy enough to test with this code (varying L1 and L2 separately).

File: src/test_nn.R (hunk 12 above)
[the remark this answers is not in the document]
Comment (aappling-usgs) [reply src/test_nn.R]: Well, i'm not actually sure this is the best place in the code to make that partition, but ideally, while we're playing with various possible hyperparameters, we would completely hold out and never look at a few years worth of test data. Then we'd only run predictions for the test data at the very very end after we've committed to a model specification. I realize this is easier said than done for lake temperature datasets, which are large but not Big in the twitter/IoT sense, but I think we should probably try because it's good practice for avoiding model overfitting / human overconfidence.

So if `test` is that completely-held-out data, then we also need some data (`dev`) that allow us to run stats like R2 as we're trying out different architectures and hyperparameters. Sometimes people split up the non-test data into different chunks on each iteration through exploring different hyperparameter values, a la k-fold cross validation. If we took that approach, these lines would be not the best place to do that division because we'd need to do the division in a different way on each iteration. But for starters we could also just designate ~70% of the non-test data as `training` and 30% as `dev`, then use `dev` rather than `test` in the places down below where `test` is currently getting used.

File: src/test_nn.R (hunk 13 above)
[the remark this answers is not in the document]
Comment (aappling-usgs) [reply src/test_nn.R]: Oh...do you actually revise the value of `training` down below? I wasn't expecting/looking for that.

[the remark this answers is not in the document]
Comment (wdwatkins) [reply src/test_nn.R]: Line 42 creates the matrix that actually is used as input features for keras.  It is a new object though, doesn't make the new `train_targets` vector necessarily

File: src/test_nn.R (hunk 9 above)
Comment (aappling-usgs) [reply src/test_nn.R]: oh, oops, i just now saw this `validation_split` argument after all that discussion up at https://github.com/USGS-R/lake-temperature-neural-networks/pull/1#discussion_r208718648. looks like keras might be holding out some data within the training set for us already?

Comment (wdwatkins) [reply src/test_nn.R]: Yes, but I think that is only used during training as an evaluation metric for when it should stop iterating.  So I think your point still stands for hyperparameter tuning.

File: src/test_nn.R (hunk 5 above)
Comment (wdwatkins) [reply src/test_nn.R]: Right.  `all_model_temps_xxx` contain all of the GLM temperature output

**Comment by wdwatkins** [comment]
fixed #2 in last commit

File: src/functions.R [hunk 15]
```diff
@@ -1,4 +1,10 @@
-identify_out_of_order <- function(vec){
+#attempting to identify bad observations (i.e. violating increasing density w/ depth)
+#This is different from cummin() in that cummin throws out everything after a deviation from 
+#monotonically increasing. When this function encounters a point that is not monotonic, it discards it,
+#but then compares to the next one. It was an attempt to not throw out entire temp profiles based on 
+#a single bad observation
+#currently not used.
```
Comment (aappling-usgs) [inline src/functions.R]: :+1:
